tiledb/storage_format/versioning/versioning.h

```diff
@@ -0,0 +1,521 @@
+   * If the versions are not compatible, an exception is raised.
+   */
+  void check_read_compatibility() {
+    check_validity();
+
+    auto lib_vsn = current_version();
+
+    // If the library version is newer or equal to the array's base version,
+    // we allow reading from the array.
+    if (lib_vsn.version_ >= version_) {
+      return;
+    }
+
+    std::string msg = "Library version " + lib_vsn.to_error_string() +
+                      " is unable to read from an array with version " +
+                      to_error_string();
+    throw std::invalid_argument(msg);
+  }
+
+  /**
+   * Check if the current library version is capable of writing to this
+   * version. If the versions are not compatible, an exception is raised.
+   */
+  void check_write_compatibility() {
+    check_validity();
+
+    auto lib_vsn = current_version();
+
+    // We allow these cases:
+    // 1. Both experimental, and exactly the same version
+    // 2. Neither experimental, and the library version is equal to or greater
+    //    than this version.
+    if (lib_vsn.is_experimental_ && is_experimental_ &&
+        lib_vsn.version_ == version_) {
+      return;
+    } else if (
+        !lib_vsn.is_experimental_ && !is_experimental_ &&
+        lib_vsn.version_ >= version_) {
+      return;
+    }
+
+    // This library version is unable to write to this disk version.
+    std::string msg = "Library version " + lib_vsn.to_error_string() +
+                      " is unable to write to an array with version " +
+                      to_error_string();
+    throw std::invalid_argument(msg);
+  }
+
+ private:
+  /* ********************************* */
+  /*         PRIVATE CONSTRUCTORS      */
+  /* ********************************* */
+
+  /**
+   * Constructor
+   *
+   * @param version The version value to use
+   * @param is_experimental Whether this is an experimental version or not.
+   */
+  constexpr format_version_t(
+      uint32_t version, bool is_experimental, bool is_valid)
+      : version_(version)
+      , is_experimental_(is_experimental)
+      , is_valid_(is_valid) {
+  }
+
+  void check_validity() const {
+    if (!is_valid_) {
+      throw std::logic_error(
+          "This is an invalid version and cannot be compared.");
+    }
+  }
+
+  /** The version number. */
+  uint32_t version_;
+
+  /** Whether this version is experimental. */
+  bool is_experimental_;
+
+  /** Whether this version is valid. */
+  bool is_valid_;
+};
+
+/**
+ * The fragment_name_version_t class exists for the single function
+ * tiledb::sm::utils::parse::get_fragment_name_version which guesses the
+ * version of a given fragment URI.
+ *
+ * BIG NOTE: This should not be confused with the version that may or may not
+ * exist in a fragment URI. This is a version of the URI itself as our URI
+ * formats have changed over time. There's at least one major issue in
+ * FragmentMetadata::get_footer_offset_and_size due to this mistake.
+ */
```

Comment:
After your change to the footer code, all I see is call to get_fragment_name_version, then seeing if the value that comes back is 1. I think we can have a function that takes in a fragment name and returns wether or not it is v1v2 as a boolean and get rid of a lot of code.